### PR TITLE
43813 hpos add missing filters on order listtable

### LINKED ISSUE #43813: [Enhancement]: HPOS Add missing filters on Order ListTable
### Describe the solution you'd like

Loading the Orders List page (/wp-admin/admin.php?page=wc-orders) on large stores with thousands/millions of orders, there are three queries that take a while to run.

**Query 1**
Created here https://github.com/woocommerce/woocommerce/blob/8.5.1/plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php#L416
```
SELECT COUNT(DISTINCT wp_wc_orders.id)
FROM wp_wc_orders
WHERE 1=1
AND (wp_wc_orders.status IN ('wc-pending','wc-processing','wc-on-hold','wc-completed','wc-cancelled','wc-refunded','wc-failed'))
AND (wp_wc_orders.type = 'shop_order')
```

**Query 2**
Created here https://github.com/woocommerce/woocommerce/blob/8.5.1/plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php#L585-L591
```
SELECT status, COUNT(*) AS cnt
FROM wp_wc_orders
WHERE type = 'shop_order'
GROUP BY status
```

**Query 3**
Created here https://github.com/woocommerce/woocommerce/blob/8.5.1/plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php#L729-L736
```
SELECT DISTINCT YEAR( t.date_created_local ) AS year, MONTH( t.date_created_local ) AS month
FROM (
SELECT DATE_ADD( date_created_gmt, INTERVAL 0 SECOND ) AS date_created_local
FROM wp_wc_orders
WHERE status != 'trash' ) t
ORDER BY year DESC, month DESC
```

Prior to HPOS, there were filters available allowing us to disable certain counts and queries from the orders list pages.

Unfortunately, not all filters currently exist on the HPOS classes, so for consistency and to allow store owners to speed up those pages, it would be great if core could add those.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by duotive**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #43813 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Since there are multiple filters added in the PR, we will test them separately.

All of the test should have HPOS authoritative.

### For `woocommerce_before_' . $this->order_type . '_list_table_view_links`
1. Add this code snippet to trigger the filter:
```php
add_filter( 'woocommerce_before_shop_order_list_table_view_links', function () {
	return array('test' => 'Filter is overriding all pages link.');
});
```
2. Go to `/wp-admin/admin.php?page=wc-orders`. Instead of link to different sub pages, you will see a line `Filter is overriding all pages link.` like so:
Screenshot 2024-01-25 at 1 54 29 PM

### For `_should_render_blank_state` filter:
0. Make sure you have at least one order in the test site.
1. Add this snippet to trigger the filter:
```php
add_filter( 'woocommerce_shop_order_list_table_should_render_blank_state', '__return_true' );
```
2. Go to `/wp-admin/admin.php?page=wc-orders`, you will see a blank order screen, as if there are no orders in the site at all.

### For `_list_table_disable_months_filter`:
1. Remove both the snippets added in the above steps to restore normal functioning.
2. Add the following snippet to disable the months filter:
```php
add_filter('woocommerce_shop_order_list_table_disable_months_filter', '__return_true');
```
3. Go to `/wp-admin/admin.php?page=wc-orders`. You will see that months filter is not added to UX any more like so:
Screenshot 2024-01-25 at 2 01 23 PM

### For `woocommerce_orders_table_query_count_sql` filter:
1. Add the following snippet to trigger, this snippet will force the order count to always return 32
```php
add_filter( 'woocommerce_orders_table_query_count_sql', function () {
	return 'SELECT 32 FROM dual';
});
```
2. Go to `/wp-admin/admin.php?page=wc-orders`., and see the order count listed. It should be 32 regardless of the number of orders there in the site like so:

Screenshot 2024-01-25 at 2 07 22 PM

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]: Let's rename this filter to add the `before` keyword, that way it will mark a clear distinction between this and ` apply_filters( "views_{$this->screen->id}", $views );` filter.
```suggestion
		$view_links = apply_filters( 'woocommerce_before_' . $this->order_type . '_list_table_view_links', $view_links );
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]: Same, slight suggestion in naming filter:
```suggestion
			'woocommerce_' . $this->order_type . '_list_table_should_render_blank_state',
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]: this should also have `$this->order_type`, either in name like you are doing above, or as a passed param.

**Review by vedanshujain — COMMENTED** [review]
Very cool changes, thanks! I have left some minor comments regarding naming of a few filters, but otherwise looks good.

**Comment by duotive** [comment]
> Very cool changes, thanks! I have left some minor comments regarding naming of a few filters, but otherwise looks good.

Thank you for reviewing them, they will be indeed helpful! I have made the changes you requested. I have noticed that an extra test failed, but I think it's unrelated. Let me know if there is anything on my end that needs to be done regarding the test.

**Comment by lsinger** [comment]
> I have noticed that an extra test failed, but I think it's unrelated.

I re-ran that check -- the test now passed. The "Build Live Branch" is expected to fail. 👍

**Comment by glagonikas** [comment]
hi @lsinger is there anything pending before this can be merged? It would be great if we could have it for 8.6.0

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php]:
```suggestion
			 * @since 8.6.0
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * @since 8.6.0
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * @since 8.6.0
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * @since 8.6.0
```

File: plugins/woocommerce/changelog/43813-hpos-add-missing-filters-on-order-listtable
Comment (lsinger) [inline plugins/woocommerce/changelog/43813-hpos-add-missing-filters-on-order-listtable]:
```suggestion
HPOS: add missing filters on order list table class.
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * Filters the list of available list table view links before the actual query runs.
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * This can be used to, e.g., remove counts from the links.
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * @param string[] $views An array of available list table view links.
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * Whether we should render a blank state so that custom count queries can be used.
```

File: plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php
Comment (lsinger) [inline plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php]:
```suggestion
		 * @param null           $should_render_blank_state `null` will use the built-in counts. Sending a boolean will short-circuit that path.
```

**Review by lsinger — CHANGES_REQUESTED** [review]
The changes requested by @vedanshujain have been addressed, so all that is left now is to replace the `@since x.x.x` tags with `8.6.0` (if this gets merged in time -- if not we can always adjust again). 

I also left a few suggestions to tweak the wording in some comments.

**Comment by duotive** [comment]
Hi @lsinger. I have made the suggested changes, please have another look.

**Review by lsinger — APPROVED** [review]
Thanks for addressing all comments, @duotive! This looks good to me. I'll go ahead and merge.

**Pull request merged**
